### ISSUE TITLE
SafeBuffer implicit type coercion?

### ISSUE THREAD

**Issue by semaperepelitsa** [rails/rails#22947]
Hello,

We are often passing around html_safe strings in our Rails app, especially in helpers. And I just discovered that appending to html_safe string results in some weird type coercion where regular string would raise a type error. Example:

``` ruby
gem "activesupport", "4.2.5"
require "active_support/core_ext/string/output_safety"

p("".html_safe + nil) #=> ""
p("".html_safe + 123) #=> "{"
p("".html_safe + Object.new) #=> "#&lt;Object:0x007fe664939778&gt;"
```

In our app we were appending nil to a string and the error was discovered only when we replaced html_safe string with regular string.

Is there any reason for this behaviour?

**Comment by rafaelfranca** [comment]
As far I know that behavior is not intentional but you would need to check. PDI!

**Comment by riseshia** [comment]
https://github.com/rails/rails/blob/7f18ea14c893cb5c9f04d4fda9661126758332b5/activesupport/lib/active_support/core_ext/string/output_safety.rb
Please check 'html_escape_interpolated_argument' method. this method escape when itself is html_safe, args is not html_safe.
I guess this is easiest way to implement '+' between SafeBuffer and String. and I think this way is natural with duck type, because object enable to respond 'to_s' could get return.

anyway, I agree to @rafaelfranca.

I think it's better to check type unless there is problem related this, to make less confusion. ;_;

**Comment by semaperepelitsa** [comment]
No news.

```
gem "activesupport", "5.1.0.beta1"
require "active_support/core_ext/string/output_safety"

p("".html_safe + nil) #=> ""
p("".html_safe + 123) #=> "{"
p("".html_safe + Object.new) #=> "#&lt;Object:0x007fe664939778&gt;"
```

**Comment by byroot** [comment]
I'd like to fix this, I have a quick draft in #42123.

The only thing I'm not sure of is wether it's OK to make this fail right away, or if we should go through some deprecation first.

I think the later. I'll finish that PR tomorrow.

**Comment by byroot** [comment]
Ok, so Rails 7.0 will throw a deprecation warning on these cases, and 7.1 will raise.
